badger/cmd/info.go

```diff
@@ -59,39 +72,129 @@ info. It also prints info about missing/extra files, and general information abo
 files (which are not referenced by the manifest).  Use this tool to report any issues about Badger
 to the Dgraph team.
 `,
-	Run: func(cmd *cobra.Command, args []string) {
-		err := printInfo(sstDir, vlogDir)
-		if err != nil {
-			fmt.Println("Error:", err.Error())
-			os.Exit(1)
-		}
-		if !opt.showTables {
-			return
-		}
-		// Open DB
-		opts := badger.DefaultOptions
-		opts.TableLoadingMode = options.MemoryMap
-		opts.Dir = sstDir
-		opts.ValueDir = vlogDir
-		opts.ReadOnly = true
-
-		db, err := badger.Open(opts)
-		if err != nil {
-			fmt.Println("Error:", err.Error())
-			os.Exit(1)
+	Run: handleInfo,
+}
+
+func handleInfo(cmd *cobra.Command, args []string) {
+	if err := printInfo(sstDir, vlogDir); err != nil {
+		fmt.Println("Error:", err.Error())
+		os.Exit(1)
+	}
+
+	// Open DB
+	opts := badger.DefaultOptions
+	opts.TableLoadingMode = options.MemoryMap
+	opts.Dir = sstDir
+	opts.ValueDir = vlogDir
+	opts.ReadOnly = true
+
+	db, err := badger.Open(opts)
+	if err != nil {
+		fmt.Println("Error:", err.Error())
+		os.Exit(1)
+	}
+	defer db.Close()
+
+	if opt.showTables {
+		tableInfo(sstDir, vlogDir, db)
+	}
+
+	if opt.showHistogram {
+		db.PrintHistogram([]byte(opt.withPrefix))
+	}
+
+	if opt.showKeys {
+		showKeys(db, opt.withPrefix)
+	}
+
+	if len(opt.keyLookup) > 0 {
+		lookup(db)
+	}
+}
+
+func showKeys(db *badger.DB, prefix string) {
+	if len(prefix) > 0 {
+		fmt.Printf("Only choosing keys with prefix: \n%s", hex.Dump([]byte(prefix)))
+	}
+	txn := db.NewTransaction(false)
+	defer txn.Discard()
+
+	iopt := badger.DefaultIteratorOptions
+	iopt.PrefetchValues = false
+	if opt.keyHistory {
+		iopt.AllVersions = true
+	}
+	it := txn.NewIterator(iopt)
+	defer it.Close()
+
+	totalKeys := 0
+	for it.Rewind(); it.Valid(); it.Next() {
+		printKey(it.Item(), false)
+		totalKeys++
+	}
+	fmt.Print("\n[Summary]\n")
+	fmt.Println("Total Number of keys:", totalKeys)
+
+}
+
+func lookup(db *badger.DB) {
+	txn := db.NewTransaction(false)
+	defer txn.Discard()
+
+	iopts := badger.DefaultIteratorOptions
+	iopts.PrefetchValues = false
+	if opt.keyHistory {
+		iopts.AllVersions = true
+		iopts.PrefetchValues = true
+	}
+	itr := txn.NewIterator(iopts)
+	defer itr.Close()
+
+	key, err := hex.DecodeString(opt.keyLookup)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
we should avoid calling `log.Fatal` because no defers will be executed. `panic` is another solution, but it is not the cleanest. Can we do better?

Reply:
We can return errors from the commands. I'll make changes so that the `info` command returns `error` to the `Root Command`. This would allow all the `defers` to be executed. 

The root command will log the error and exit
https://github.com/dgraph-io/badger/blob/ff2b7fe72292c0c971de7c8659b13542aa71a5c0/badger/cmd/root.go#L39-L44

Sounds good?

Reply:
yeah, sounds better.

Reply:
Fixed in https://github.com/dgraph-io/badger/commit/0b9bab5832d6a7068ade67fd93fd0e1bdd70547d